### ISSUE TITLE
Using 3.0 with external react and external rendering

### ISSUE THREAD

**Issue by backjo** [gaearon/react-hot-loader#500]
I have an application where React, ReactDOM, and the rendering of any components are all outside of my webpack build. Basically, webpack will build a bundle that exposes each component via a require() call. The non-webpack application will dynamically require these components and handle rendering them with ReactDOM.render().

In version 1, every one of these components would have the following snippet inside of their componentDidMount method, which would make hot reloading work. 

```js
if (module.hot) {
  let self = this;
  require('react-hot-loader/Injection').RootInstanceProvider.injectProvider({
    getRootInstances: function () {
      return [self];
    }
  });
}
```

So, to summarize the setup: Application A lives outside of webpack (and owns the React/ReactDOM libraries), while application B is in webpack (with React and ReactDOM as externals). B's top-level components are exposed to A's module loader, which will dynamically load them at runtime and handle rendering them. 

Any thoughts on how this might work with v3? From the documentation, I couldn't see a way to handle this since I don't control the top level render inside of the webpack application.

**Comment by SystemParadox** [comment]
I started upgrading to v3 and ran into a similar issue with our popup system. When we call `renderPopup(<FooPopup />)`, it adds a separate `div` with its own react tree. The popups are totally independent - the original page retains no state to say that the popup is open, so re-rendering the page does not re-render the popup.

This works fine with v1, but I can't think how to make this work with v3 when the component is imported in one module (the page), and rendered in another (the popup manager).

**Comment by neoziro** [comment]
I am sorry, external rendering is not supported in v3.

**Comment by Aghassi** [comment]
@theKashey Sorry to post on a closed post, but will this issue been addressed in the future since it use to be supported? I've looked through the docs and commits and don't seem to see anything. If it isn't supported, is there a recommended alternative for users who require this functionality?

**Comment by theKashey** [comment]
@Aghassi - sorry, but this problem is absolutely unclear for me. Nor what you want to do, nor why it does not work.

**Comment by Aghassi** [comment]
@theKashey Sure, let me try and define the problem set.

In https://github.com/gaearon/react-hot-loader/blob/v1.0.6/CHANGELOG.md#106 there was a change that allowed you to specify the root component in environments where React is standalone (this means external to the webpack bundle, or loaded by some other entity). In our use case, we have a Require base system that loads our webpack bundle into the browser as well as React. This means that the bundle is not in charge of knowing the render cycle for something like a hot reload. If we use hot reload as is, the server (which is not local, it is remote) will not know how to respond to the reload request, resulting in nothing happening.
Essentially, I'm looking for a way to remedy this given that v3 and v4 don't seem to support the `Injection` class anymore. The method that the original author provided is similar to what we used, hence why I posted in this thread.

**Comment by theKashey** [comment]
@Aghassi - is sounds like you need this function - https://github.com/gaearon/react-hot-loader/blob/master/src/reactHotLoader.js#L82
```js
import ReactHotLoader from 'react-hot-loader';
ReactHotLoader.patch(AnythingDucksAsReact); // land RHL patches where you want.
```

**Comment by Aghassi** [comment]
@theKashey That looks promising! I assume that goes in the file that is bundled and I just do the normal hot(module)(component) syntax? Or is there more to it? Documentation is a bit sparse (I’m happy to contribute back my findings so others can leverage it).

**Comment by theKashey** [comment]
RHL needs 2 things:
- "React" to patch
- "Module" to accept

You can provide your own version of react to patch - just `ReactHotLoader.patch` it
You are always giving your "module" to `hot(module)`

I am not sure why you have to manually patch react, as long all the modules are importing react in the same way - `import React from 'react'`.
RHL v4 does not rely on any React internals - only on React.createElement, and if something is not working out of the box - the right question is "why".
